backend-shared/util/manager.go

```diff
@@ -0,0 +1,126 @@
+		return nil, fmt.Errorf("error adding apis.kcp.dev/v1alpha1 to scheme: %w", err)
+	}
+
+	apiExportClient, err := client.New(cfg, client.Options{Scheme: scheme})
+	if err != nil {
+		return nil, fmt.Errorf("error creating APIExport client: %w", err)
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create discovery client: %w", err)
+	}
+
+	return getControllerManager(ctx, cfg, apiExportClient, discoveryClient, log, apiExportName, opts)
+}
+
+// getControllerManager returns a standard controller-runtime manager in a non-KCP environment and a KCP workspace aware manager in the presence of KCP APIs
+func getControllerManager(ctx context.Context, restConfig *rest.Config, apiExportClient client.Client, discoveryClient discovery.DiscoveryInterface, log *logr.Logger, apiExportName string, opts ctrl.Options) (ctrl.Manager, error) {
+	var mgr ctrl.Manager
+
+	isKCPEnvironment, err := kcpAPIsGroupPresent(discoveryClient)
+	if err != nil {
+		return nil, err
+	}
+
+	if isKCPEnvironment {
+		log.Info("Looking up virtual workspace URL")
+		cfg, err := restConfigForAPIExport(ctx, restConfig, apiExportClient, apiExportName)
+		if err != nil {
+			log.Error(err, "error looking up virtual workspace URL")
+		}
+
+		log.Info("Using virtual workspace URL", "url", cfg.Host)
+
+		opts.LeaderElectionConfig = restConfig
+		mgr, err = kcp.NewClusterAwareManager(cfg, opts)
+		if err != nil {
+			return nil, fmt.Errorf("unable to start cluster aware manager: %w", err)
+		}
+	} else {
+		log.Info("The apis.kcp.dev group is not present - creating standard manager")
+		mgr, err = ctrl.NewManager(restConfig, opts)
+		if err != nil {
+			return nil, fmt.Errorf("unable to start manager: %w", err)
+		}
+	}
+
+	return mgr, nil
+}
+
+// restConfigForAPIExport returns a *rest.Config properly configured to communicate with the endpoint for the
+// APIExport's virtual workspace.
+func restConfigForAPIExport(ctx context.Context, cfg *rest.Config, apiExportClient client.Client, apiExportName string) (*rest.Config, error) {
+
+	var apiExport apisv1alpha1.APIExport
+
+	if apiExportName != "" {
+		if err := apiExportClient.Get(ctx, types.NamespacedName{Name: apiExportName}, &apiExport); err != nil {
+			return nil, fmt.Errorf("error getting APIExport %q: %w", apiExportName, err)
+		}
+	} else {
+		exports := &apisv1alpha1.APIExportList{}
+		if err := apiExportClient.List(ctx, exports); err != nil {
+			return nil, fmt.Errorf("error listing APIExports: %w", err)
+		}
+		if len(exports.Items) == 0 {
+			return nil, fmt.Errorf("no APIExport found")
+		}
+		if len(exports.Items) > 1 {
+			return nil, fmt.Errorf("more than one APIExport found")
+		}
+		apiExport = exports.Items[0]
+	}
+
+	if len(apiExport.Status.VirtualWorkspaces) < 1 {
+		return nil, fmt.Errorf("APIExport %q status.virtualWorkspaces is empty", apiExportName)
+	}
+
+	cfg = rest.CopyConfig(cfg)
+	// TODO(ncdc): sharding support
```

Comment:
We can open a story for this item, then convert this TODO into the format used by the GitOps Service: `// TODO: GITOPSRVCE-XXX - implement sharding of virtual workspaces`, where `XXX` is the story id.

Reply:
I've copied the TODO comment from the [controller-runtime example](https://github.com/kcp-dev/controller-runtime-example/blob/813d765af3a26113c1cd43101544cb763e24f2b5/main.go#L200). 

Created https://issues.redhat.com/browse/GITOPSRVCE-204 to track it on our board.

Since sharding is not yet supported in the KCP package I haven't added the implementation details. Please feel free to update the story